File: src/main/groovy/com/bugsnag/android/gradle/BugsnagPluginExtension.java
package com.bugsnag.android.gradle;

import java.util.Map;

/**
 * Defines configuration options (Gradle plugin extensions) for the BugsnagPlugin
 */
public class BugsnagPluginExtension {

    String endpoint = "https://upload.bugsnag.com";
    String releasesEndpoint = "https://build.bugsnag.com";

    boolean autoUpload = true;
    boolean autoReportBuilds = true;
    boolean uploadDebugBuildMappings = false;
    boolean overwrite = false;
    int retryCount = 0;
    Boolean ndk = null;
    String sharedObjectPath = null;
    String projectRoot = null;
    boolean failOnUploadError = true;
    int requestTimeoutMs = 60000;

    // release API values
    String builderName = null;
    Map<String, String> metadata = null;
    Map<String, String> objdumpPaths = null;

    public String getEndpoint() {
        return endpoint;
    }

    public void setEndpoint(String endpoint) {
        this.endpoint = endpoint;
    }

    public String getReleasesEndpoint() {
        return releasesEndpoint;
    }

    public void setReleasesEndpoint(String releasesEndpoint) {
        this.releasesEndpoint = releasesEndpoint;
    }

    public boolean isAutoUpload() {
        return autoUpload;
    }

    public void setAutoUpload(boolean autoUpload) {
        this.autoUpload = autoUpload;
    }

    public boolean isAutoReportBuilds() {
        return autoReportBuilds;
    }

    public void setAutoReportBuilds(boolean autoReportBuilds) {
        this.autoReportBuilds = autoReportBuilds;
    }

    public boolean isUploadDebugBuildMappings() {
        return uploadDebugBuildMappings;
    }

    public void setUploadDebugBuildMappings(boolean uploadDebugBuildMappings) {
        this.uploadDebugBuildMappings = uploadDebugBuildMappings;
    }

    public boolean isOverwrite() {
        return overwrite;
    }

    public void setOverwrite(boolean overwrite) {
        this.overwrite = overwrite;
    }

    public int getRetryCount() {
        return retryCount;
    }

    public void setRetryCount(int retryCount) {
        this.retryCount = retryCount;
    }

    public Boolean getNdk() {
        return ndk;
    }

    public void setNdk(Boolean ndk) {
        this.ndk = ndk;
    }

    public String getSharedObjectPath() {
        return sharedObjectPath;
    }

    public void setSharedObjectPath(String sharedObjectPath) {
        this.sharedObjectPath = sharedObjectPath;
    }

    public String getProjectRoot() {
        return projectRoot;
    }

    public void setProjectRoot(String projectRoot) {
        this.projectRoot = projectRoot;
    }

    public boolean isFailOnUploadError() {
        return failOnUploadError;
    }

    public void setFailOnUploadError(boolean failOnUploadError) {
        this.failOnUploadError = failOnUploadError;
    }

    public int getRequestTimeoutMs() {
        return requestTimeoutMs;
    }

    public void setRequestTimeoutMs(int requestTimeoutMs) {
        this.requestTimeoutMs = requestTimeoutMs;
    }

    public String getBuilderName() {
        return builderName;
    }

    public void setBuilderName(String builderName) {
        this.builderName = builderName;
    }

    public Map<String, String> getMetadata() {
        return metadata;
    }

    public void setMetadata(Map<String, String> metadata) {
        this.metadata = metadata;
    }

    public Map<String, String> getObjdumpPaths() {
        return objdumpPaths;
    }

    public void setObjdumpPaths(Map<String, String> objdumpPaths) {
        this.objdumpPaths = objdumpPaths;
    }
}

File: src/main/kotlin/com.bugsnag.android.gradle/BugsnagPluginExtension.kt
package com.bugsnag.android.gradle

import groovy.lang.Closure
import org.gradle.api.Action
import org.gradle.api.model.ObjectFactory
import org.gradle.api.provider.ListProperty
import org.gradle.api.provider.MapProperty
import org.gradle.api.provider.Property
import org.gradle.util.ConfigureUtil
import java.io.File

// To make kotlin happy with gradle's nullability
private val NULL_STRING: String? = null
private val NULL_BOOLEAN: Boolean? = null

/**
 * Defines configuration options (Gradle plugin extensions) for the BugsnagPlugin
 */
// After Gradle 5.2, this can use service injection for injecting ObjectFactory
open class BugsnagPluginExtension(objects: ObjectFactory) {

    val sourceControl: SourceControl = objects.newInstance()

    val enabled: Property<Boolean> = objects.property<Boolean>()
        .convention(true)

    val uploadJvmMappings: Property<Boolean> = objects.property<Boolean>()
        .convention(true)

    val uploadNdkMappings: Property<Boolean> = objects.property<Boolean>()
        .convention(NULL_BOOLEAN)

    val reportBuilds: Property<Boolean> = objects.property<Boolean>()
        .convention(true)

    val uploadDebugBuildMappings: Property<Boolean> = objects.property<Boolean>()
        .convention(false)

    val autoUpdateBuildUuid: Property<Boolean> = objects.property<Boolean>()
        .convention(true)

    val endpoint: Property<String> = objects.property<String>()
        .convention("https://upload.bugsnag.com")

    val releasesEndpoint = objects.property<String>()
        .convention("https://build.bugsnag.com")

    val overwrite: Property<Boolean> = objects.property<Boolean>()
        .convention(false)

    val retryCount: Property<Int> = objects.property<Int>()
        .convention(0)

    val sharedObjectPaths: ListProperty<File> = objects.listProperty<File>()
        .convention(emptyList())

    val projectRoot: Property<String> = objects.property<String>().convention(NULL_STRING)

    val failOnUploadError: Property<Boolean> = objects.property<Boolean>()
        .convention(true)

    val requestTimeoutMs: Property<Long> = objects.property<Long>()
        .convention(60000)

    // release API values
    val builderName: Property<String> = objects.property<String>().convention(NULL_STRING)

    val metadata: MapProperty<String, String> = objects.mapProperty<String, String>()
        .convention(emptyMap())

    val objdumpPaths: MapProperty<String, String> = objects.mapProperty<String, String>()
        .convention(emptyMap())

    // exposes sourceControl as a nested object on the extension,
    // see https://docs.gradle.org/current/userguide/custom_gradle_types.html#nested_objects

    fun sourceControl(closure: Closure<SourceControl>) {
        ConfigureUtil.configure(closure, sourceControl)
    }

    fun sourceControl(action: Action<SourceControl>) {
        action.execute(sourceControl)
    }
}
